Library/Homebrew/utils/update-completions.zsh

```diff
@@ -0,0 +1,48 @@
+#!/bin/zsh
+
+typeset -U PATH path
+path+=( $HOMEBREW_REPOSITORY/bin ) > /dev/null
+
+autoload -Uz compinit && compinit
+
+zmodload -i zsh/zutil
+zstyle ':completion:*' use-cache true
+zstyle ":completion:*" cache-policy __brew_completion_caching_policy
+
+__brew_completion_caching_policy() {
+  local -a tmp
+
+  # invalidate if cache file is missing or >=2 weeks old
+  tmp=( $1(mw-2N) )
+  (( $#tmp )) || return 0
+
+  # otherwise, invalidate if latest tap index file is missing or newer than cache file
+  tmp=( ${HOMEBREW_REPOSITORY:-/usr/local/Homebrew}/Library/Taps/*/*/.git/index(om[1]N) )
+  [[ -z $tmp || $tmp -nt $1 ]]
+}
+
+local comp_cachename
+local -a list
+
+comp_cachename=brew_all_commands
+if _cache_invalid $comp_cachename; then
+  HOMEBREW_CACHE=$(brew --cache)
+  HOMEBREW_REPOSITORY=$(brew --repo)
+  [[ -f "$HOMEBREW_CACHE/all_commands_list.txt" ]] &&
+    list=($(cat "$HOMEBREW_CACHE/all_commands_list.txt")) ||
+    list=($(cat "$HOMEBREW_REPOSITORY/completions/internal_commands_list.txt"))
+  list=(${list:#*instal}) # Exclude instal, uninstal, etc.
+  _store_cache $comp_cachename list
+fi
+
+comp_cachename=brew_formulae
+if _cache_invalid $comp_cachename; then
+  list=( $(brew search) )
```

Comment:
What about having the output of this be cached in `HOMEBREW_CACHE` like `all_commands_list.txt` on `brew update` instead? That would enable it to be used in Bash completion too. CC @alebcay for thoughts.

Reply:
`__brew_installed_formulae` and `__brew_installed_casks` could also be cached. Those are quite slow, too.

Reply:
> I think ideally the formula list would be generated with Ruby and cached in `HOMEBREW_CACHE` and then consumed by whatever completions we'd like. We could still store a copy of this list in the Zsh completions cache if that makes the completion execution marginally faster.

The Zsh completion code could just read it straight out of `HOMEBREW_CACHE`, instead of using Zsh's own caching mechanism. I need to test it, but I think it should be equally fast.

Reply:
> Homebrew's Zsh completion code could just read the completions straight out of HOMEBREW_CACHE

I believe this is what we're doing right now after #7767 was merged. I left the existing completion cache code in since it would still be useful for the other Zsh completions (formulae and casks), and wouldn't hurt anything.

Reply:
Ah, you're right: https://github.com/Homebrew/brew/pull/7767/files#diff-c0df35ca56e49697ec906a8ce9a45448R151

Sorry, I missed that.

Reply:
> What about having the output of this be cached in `HOMEBREW_CACHE` like `all_commands_list.txt` on brew update instead? That would enable it to be used in Bash completion too. 

@MikeMcQuaid I'm all for it, but can we do that as a separate PR? I would like to get this one merged now.

Reply:
@marlonrichert I'd rather not make the `cmd/update.sh` change given the discussion above, sorry. I'm open to other parts of this PR being merged. When `cmd/update.sh` fails: no-one can update Homebrew so I'm extremely reluctant to make changes there that aren't 100% necessary.

Reply:
> `__brew_installed_formulae` and `__brew_installed_casks` could also be cached. Those are quite slow, too.

This seems like a good idea. This cache could be updated on install/uninstall of these commands.

Reply:
> I'd rather not make the `cmd/update.sh` change given the discussion above, sorry. I'm open to other parts of this PR being merged. When `cmd/update.sh` fails: no-one can update Homebrew so I'm extremely reluctant to make changes there that aren't 100% necessary.

OK, I'll remove the `cmd/update.sh` changes from this PR.

Reply:
Thanks! Please roll Library/Homebrew/utils/update-completions.zsh back into completions/zsh/_brew too. After that, though, I'm not sure what's actually changing here?

Reply:
What’s left are three things that speed up Zsh completions:
* Enable Zsh completion caching for `brew` arguments by default. (Right now, it’s disabled by default.)
* Use `$(<file)` instead of `$(cat file). (The former is faster than the latter, because it does not spawn a new process, which the latter does.)
* Don’t call `_cache_invalid` before calling `_retrieve_cache`. (The latter already calls the former. No need to call it twice.)

Reply:
Ok, cool, sounds good 👍

Reply:
@MikeMcQuaid  I made the changes. Please review.